text/2021-02-24-coprocessor-plugin.md

```diff
@@ -0,0 +1,161 @@
+# Coprocessor Plugin
+
+## Summary
+
+Add a general and pluggable coprocessor framework for RawKV mode.
+
+## Motivation
+
+TiKV is the storage component in the TiDB ecosystem, however, the distribution computation principle suggests that computation should be as close to the data source as possible. Therefore, TiKV has embedded a subset of the TiDB executor framework to push down some computation tasks when applicable.
+
+But TiKV's capability should be far beyond that, as many distributed components can be built on top of TiKV, such as cache, full text search engine, graph database and NoSQL database. And same as TiDB, these product will also like to push down specific computation to TiKV, which requires the coprocessor to be customizable, aka pluggable.
+
+For instance, a full-text seraching engine will persist the origin document and n-gram index on TiKV. It'll be a waste of resource if we read back and then update the index from a client. In contrary, the coprocessor plugin can generate the index from the origin document, and update the index inplace. What's more, the coprocessor plugin can perform index scan directly on TiKV.
+
+The goals of the coprocessor plugin are:
+
+- Do what client can do (on single region)
```

Comment:
Not necessary limited to one region. Though should be simple to start with single region first.